First version of stack deployment tutorial

aanand opened the pull request:
Based on @vdemeester's work in #672 

To try this tutorial out, you need a 1.13 daemon, a 1.13 client and a [master build of Compose](https://dl.bintray.com/docker-compose/master/).

johndmulhausen commented:
Deploy preview ready!

Built with commit 0ca615c46cf50b9a8f52e53764d3954308ec67a6

https://deploy-preview-685--docker-docs.netlify.com

dnephin commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,250 @@
+---
+
+When running Docker Engine in swarm mode, you can use `docker stack deploy`
+(or just `docker deploy`) to deploy a complete application stack to the swarm.
```
We're planning on keeping `docker deploy` experimental in 1.13 so that we can change it later if we need to.  We might want to avoid calling it out here since it won't be available to everyone like `docker stack deploy`.

dnephin commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,250 @@
+The `deploy` command accepts a stack description in the form of a [Compose
+file](/compose/compose-file.md).
+
+The `docker deploy` command supports any Compose file of version "3.0" or
```
`docker stack deploy`

dnephin commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,250 @@
+    Compose doesn't take advantage of swarm mode, and deploys everything to a
+    single node. You can safely ignore this.
+
+    <!-- TODO: this warning needs to be updated for `docker deploy` -->
```
I think this is done: https://github.com/docker/compose/pull/4170/ (needs merged)

Unless you mean something else

dnephin commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,250 @@
+
+1.  Create the stack with `docker deploy`:
+
+        $ docker deploy --compose-file docker-compose.yml stackdemo
```
`docker stack deploy`

dnephin commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,250 @@
+    The last argument is a name for the stack. Each network, volume and service
+    name is prefixed with the stack name.
+
+2.  Check that it's running with `docker service ls`:
```
`docker stack services stackdemo` might be more appropriate here?

dnephin commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,250 @@
+        Removing service stackdemo_redis
+        Removing network stackdemo_default
+
+4.  Bring the registry down ith `docker service rm`:
```
`s/ith/with/`

aanand commented:
@dnephin Thanks! Addressed.

mstanleyjones commented:
Oops, please rebase your work on `vnext-engine` branch since it requires 1.13.

mstanleyjones commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,242 @@
+a stack description in the form of a [Compose file](/compose/compose-file.md).
+
+The `docker stack deploy` command supports any Compose file of version "3.0" or
+above. If you have an older version, upgrading is simple - see the
```
Let's not editorialize about how simple it is. :) Something like "If you have an older version, see <link>."

mstanleyjones commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,242 @@
+Because a swarm consists of multiple Docker Engines, a registry is required to
+distribute images to all of them. You can use the
+[Docker Hub](https://hub.docker.com), or maintain your own, but for the purposes
+of this tutorial it's sufficient to create a throwaway registry.
```
change "it's sufficient to create a throwaway registry" to something like "This tutorials creates a simple registry, which you can discard afterward."

mstanleyjones commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,242 @@
+
+        $ docker service create --name registry --publish 5000:5000 registry:2
+
+2.  Check its status with `docker service ls`:
```
Check the status of the registry

mstanleyjones commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,242 @@
+of this tutorial it's sufficient to create a throwaway registry.
+
+1.  Start the registry as a service on your swarm:
+
```
Please use fenced code blocks instead of relying on indentation.

mstanleyjones commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,242 @@
+2.  Check its status with `docker service ls`:
+
+        $ docker service ls
+        ID            NAME      REPLICAS  IMAGE                                                                               COMMAND
```
As above, fenced code blocks. Please separate the command input from the output with a carriage return.

mstanleyjones commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,242 @@
+
+2.  Create a file called `app.py` in the project directory and paste this in:
+
+        from flask import Flask
```
When you change this to a fenced version, use Dockerfile as the highlighter.

```markdown
```Dockerfile
From flask import Flask
```

mstanleyjones commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,242 @@
+        RUN pip install -r requirements.txt
+        CMD ["python", "app.py"]
+
+5.  Finally, create a file called `docker-compose.yml` and paste this in:
```
s/Finally,//

mstanleyjones commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,242 @@
+
+5.  Finally, create a file called `docker-compose.yml` and paste this in:
+
+        version: '3'
```
Use fences for this code block and designate it as `yaml` in the highlighter.

mstanleyjones commented on engine/swarm/stack-deploy.md:
```diff
@@ -0,0 +1,242 @@
+        Creating stackdemo_redis_1
+        Creating stackdemo_web_1
+
+2.  Check that it's running with `docker-compose ps`:
```
Check that the app is running...

aanand commented:
Rebased on `vnext-engine` and addressed @mstanleyjones' comments. I don't have permission to view the build result, so I don't know why it's failing, sorry!

aanand commented:
By the way, this tutorial also depends on Compose version 1.10, which isn't out yet. I don't know if that affects what branch it should be rebased on.

mstanleyjones commented:
yeah, sorry that you can't see the Netlify build failure logs. Netlify is working to make this feature available. I'll have a look and let you know what's going on.

We should probably add a milestone for Compose 1.10 and tag this into that milestone as well. As far as the branches, there is a `vnext-compose` but let's keep it in `vnext-engine` for now, and we can work onto getting it into all the places it needs to go.

cc/ @londoncalling can you please create the milestone and tag this into it?

mstanleyjones commented:
Also the Netlify check doesn't apply to `vnext-engine` branch right now, so ignore its failure. I had to change the base in the pull request in Github too (the Edit button next to the PR title). Now it is correctly only showing one commit in the PR.

mstanleyjones commented on engine/swarm/stack-deploy.md:
````diff
@@ -0,0 +1,296 @@
+
+1.  Start the registry as a service on your swarm:
+
+    ```
````
Please add the `bash` highlighter after the fence. like this:

```markdown
```bash
```

Also add it to the other codeblocks that are terminal commands. There are lots of other highlighters such as `markdown`, `dockerfile`, `json`, `yaml`, `go`, etc.

mstanleyjones commented on engine/swarm/stack-deploy.md:
````diff
@@ -0,0 +1,296 @@
+
+3.  Create a file called `requirements.txt` and paste these two lines in:
+
+    ```
````
For this one you can use the `none` highlighter.

aanand commented:
Tagged all the code blocks with the appropriate highlighter, per @mstanleyjones' advice
